### PR TITLE
feat: Add tests for mobile experience

### PR DESCRIPTION
**Pull request by avattipalli**
## What/Why?
As part of this change, we are adding a new test that can verify checkout experience on ios mobile rendering. Along with it, we need to restructure tests in a way that to differentiate between tests that runs on desktop configuration and tests that run on mobile emulations.

## Testing
Screenshot 2024-02-26 at 5 00 21 PM

### CODE REVIEW

File: packages/functional/pages/checkout-page.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export const PROVINCE_CODE_SELECT = '[data-test="provinceCodeInput-select"]';
+export const POSTCODE_INPUT = '[data-test="postCodeInput-text"]';
+export const CHECKOUT_SHIPPING_CONTINUE = '#checkout-shipping-continue';
+export const BIGPAY_CC_NUMBER_IFRAME = '#bigpaypay-ccNumber iframe';
+export const BIGPAY_CC_EXPIRY_IFRAME = '#bigpaypay-ccExpiry iframe';
+export const BIGPAY_CC_NAME_IFRAME = '#bigpaypay-ccName iframe';
+export const BIGPAY_CC_CVV_IFRAME = '#bigpaypay-ccCvv iframe';
+export const ORDER_CONFIRMATION_HEADING = '[data-test="order-confirmation-heading"]';
+
+export * as CheckoutPage from './checkout-page';
```
Comment (chanceaclark) [inline packages/functional/pages/checkout-page.ts:20]: See my previous comment but we should prefer [user-facing attributes instead XPath or CSS selectors](https://playwright.dev/docs/best-practices#prefer-user-facing-attributes-to-xpath-or-css-selectors). I would just inline stuff, e.g. `await page.getByRole('textbox', { name: 'email' }).fill('example@example.com')`.

File: packages/functional/actions/checkout-actions.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import { faker } from '@faker-js/faker';
+import { expect, Page } from '@playwright/test';
+
+import { CheckoutPage } from '../pages/checkout-page';
+
+const testUser = faker.person.firstName();
+
+export async function fillGuestCustomerEmailAndShipping(page: Page) {
+  await page.locator(CheckoutPage.GUEST_CUSTOMER_EMAIL).click();
```
[the remark this answers is not in the document]
Comment (avattipalli) [reply packages/functional/actions/checkout-actions.ts:9]: I removed most of them but looks like I forgot a couple of places. Goingto remove them.

File: packages/functional/actions/checkout-actions.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  await page.locator(CheckoutPage.GUEST_CUSTOMER_EMAIL).fill(faker.internet.email());
```
[the remark this answers is not in the document]
Comment (avattipalli) [reply packages/functional/actions/checkout-actions.ts:10]: I was aiming for writing I18N tests for shipping related tests in the near future and I'm not sure if we use test based identifiers for such cases. I agree that in majoirity of cases, we should be testing like an end user but this might be a special case considering for I18N testing.

File: packages/functional/tests/ui/mobile/checkout-experience.spec.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+test('Checkout experience on ios mobile', async ({ page }) => {
+  await ProductActions.addProductToCart(page, sampleProduct);
+
+  await page.getByRole('link', { name: 'Cart Items 1' }).click();
+  await expect(page.getByRole('heading', { level: 1, name: 'Your cart' })).toBeVisible();
+
+  await page.getByRole('link', { name: 'Proceed to checkout' }).click();
+
+  const testUser = await CheckoutActions.fillGuestCustomerEmailAndShipping(page);
```
[the remark this answers is not in the document]
Comment (avattipalli) [reply packages/functional/tests/ui/mobile/checkout-experience.spec.ts:31]: We can but I was looking to write more tests real soon that will revisit shipping and billing.

File: packages/functional/actions/checkout-actions.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (chanceaclark) [reply packages/functional/actions/checkout-actions.ts:10]: You'd do something like this:
```
await page.goto('https://catalyst.dev/es/checkout');
await page.getByRole('textbox', { name: 'correo electrónico' });
```

I feel like those would be separate test cases, but we should aim to inline things as much as possible and then abstract whenever we see a consistent pattern. We will probably have a few tests that we fill out the shipping information differently than this method does, specifically i18n tests for other countries might include additional fields.

[the remark this answers is not in the document]
Comment (avattipalli) [reply packages/functional/actions/checkout-actions.ts:10]: I can move that into inline but ideally we should be able to run all tests against different locales, don't you think so? 

If we have to have do
`await page.getByRole('textbox', { name: 'correo electrónico' });` for Spanish and
`await page.getByRole('textbox', { name: 'email' });` for English then we end up writing way more tests don't we? 

Even if we can manage those strings in a JSON file for different locales and load those fixtures while running specific tests where we have to consider i18n, their maintenance will be difficult when we move fast. We will have to update across many places for different locales when a single text based selector changes in such cases.

[the remark this answers is not in the document]
Comment (chanceaclark) [reply packages/functional/actions/checkout-actions.ts:10]: >we end up writing way more tests don't we?

Won't we still have write individual tests for each locale anyways?

>Even if we can manage those strings in a JSON file for different locales and load those fixtures while running specific tests where we have to consider i18n, their maintenance will be difficult when we move fast. We will have to update across many places for different locales when a single text based selector changes in such cases.

My big concern here is that we are abstracting this too early. We don't even have any tests written for i18n yet, so we don't even know what each test should do or if there are any edge-cases. I personally believe that we should just duplicate test logic into each test and then come back to abstracting it when we truly believe there is a need to do so.

[the remark this answers is not in the document]
Comment (avattipalli) [reply packages/functional/actions/checkout-actions.ts]: > Won't we still have write individual tests for each locale anyways?
Going to keep that discussion for next time when we work on locale tests :)

> My big concern here is that we are abstracting...
I updated the test to confine every action to stay within the test case since we aren't reusing it for now. I also updated almost all locators be identify as a end user. In two cases I had to rely on test locators since the ios rendering doesn't show role based selectors for those. Might be something we can fix later?

**Comment by avattipalli** [comment]
@chanceaclark I had to skip the test for now since this is failing to install the required configuration on github workflows. Test works fine locally. I need to setup a new job that runs similar tests on a mac platform so that the necessary ios configuration is available to test these.

File: packages/functional/tests/ui/mobile/checkout-experience.spec.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  await page.locator(CheckoutPage.PROVINCE_CODE_SELECT).selectOption('MA');
+  await page.getByText('Postal Code').fill('01762');
+  await expect(page.locator('#checkout-shipping-continue')).toContainText('Continue');
+
+  await page.getByRole('button', { name: 'Continue' }).click();
+
+  await expect(page.getByRole('heading', { name: 'Payment', exact: true })).toBeVisible();
+
+  await page
+    .frameLocator(CheckoutPage.BIGPAY_CC_NUMBER_IFRAME)
```
Comment (chanceaclark) [inline packages/functional/tests/ui/mobile/checkout-experience.spec.ts:49]: I'm curious if we can remove the `frameLocaltor` here? I assume not, but would love to see if it's possible.

File: packages/functional/tests/ui/mobile/checkout-experience.spec.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    .fill(testUser);
+
+  await page.frameLocator(CheckoutPage.BIGPAY_CC_CVV_IFRAME).getByLabel('CVV').fill('211');
+  await page.getByRole('button', { name: 'Place Order' }).click();
+  await expect(page.locator(CheckoutPage.ORDER_CONFIRMATION_HEADING)).toBeVisible({
+    timeout: 15000,
+  });
+  await expect(page.locator(CheckoutPage.ORDER_CONFIRMATION_HEADING)).toContainText(
+    `Thank you ${testUser}!`,
+  );
```
Comment (chanceaclark) [inline packages/functional/tests/ui/mobile/checkout-experience.spec.ts:70]: Could we do something like this instead?
```ts
await page.waitForLoadState('networkidle');
await expect(
  page.getByRole('heading', { name: `Thank you ${testUser}!`, level: 1 }),
).toBeVisible();
```

**Review by chanceaclark — COMMENTED** [review]
Few comments that maybe unblock us from not having to create another runner for this.

FYI, if I don't know an elements `role` or `name`, in devtools in the browser you can select and element and see this info on the accessibility tab:
Screenshot 2024-03-01 at 12 10 44

File: packages/functional/tests/ui/mobile/checkout-experience.spec.ts (hunk 5 above)
Comment (avattipalli) [reply packages/functional/tests/ui/mobile/checkout-experience.spec.ts:53]: Tried this but it wasn't able to locate the elements without the frameLocator.

File: packages/functional/tests/ui/mobile/checkout-experience.spec.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+  await expect(
+    page.getByRole('link', { name: '[Sample] Able Brewing System' }).first(),
+  ).toBeVisible();
+});
+
+/**
+ * @todo skipping this test as we need to setup a new gh action for ios tests that run on mac config
+ *
+ */
```
[the remark this answers is not in the document]
Comment (avattipalli) [reply packages/functional/tests/ui/mobile/checkout-experience.spec.ts:25]: I knew we could do this but I was trying to run it with configuration as close as possible to a mobile rendering oppossed to setting a viewport. It's probably best for now to just go with this until we have more tests that might require additional setup.

**Review by chanceaclark — APPROVED** [review]
Overall I think it looks good, but maybe just remove the `@todo` comment.
